src/quo2/components/notifications/activity_log/view.cljs

```diff
@@ -155,18 +116,42 @@
     :style               style/unread-dot-container}
    [rn/view {:style style/unread-dot}]])
 
+(defmulti footer-item-view (fn [item _ _] (:type item)))
+
+(defmethod footer-item-view :button
+  [{:keys [label subtype disable-when] :as button} replying? reply-input]
+  (let [size         (if replying? 40 24)
+        common-style (when replying?
+                       {:padding-vertical 9
+                        :flex-grow        1
+                        :flex-basis       0})]
+    [button/button
+     (-> button
+         (assoc :size size)
+         (assoc :type subtype)
+         (assoc :disabled (and replying? (disable-when @reply-input)))
+         (update :style merge common-style {:margin-right 8}))
+     label]))
+
+(defmethod footer-item-view :status
+  [{:keys [label subtype]} _ _]
+  [status-tags/status-tag
+   {:size   :small
+    :label  label
+    :status {:type subtype}}])
+
 (defn- footer
-  [_]
+  [_ _]
   (let [reply-input (reagent/atom "")]
-    (fn [{:keys [replying? on-update-reply status button-1 button-2]}]
+    (fn [timestamp {:keys [replying? items] :as props}]
       [:<>
        (when replying?
-         [activity-reply-text-input reply-input on-update-reply])
-       (cond (some? status)
-             [activity-status status]
-
-             (or button-1 button-2)
-             [activity-buttons button-1 button-2 replying? reply-input])])))
+         [activity-reply-text-input props reply-input])
+       (when items
+         [rn/view style/footer-container
+          (for [[i item] (map-indexed vector items)]
+            ^{:key (str "footer-item-" timestamp "-" i)}
```

Comment:
Any reason why you're adding fixed strings to build the key? I see other parts of the codebase doing this, but I've never seen the utility. The key doesn't need to be human readable. Just `^{:key (str timestamp i)}` is enough. If you're sure the items are different (which I think they'll always be), you can use `{:key (hash item)}` too.

Reply:
Calculating hash for each item as a key doesn't sound like a good idea. But agree that "footer-item-" prefix is a bit redundant even though I don't see any problem with that since we do concatenation anyway.

Reply:
There's no problem per se, it's just pointless because keys don't need to be human readable.

>Calculating hash for each item as a key doesn't sound like a good idea.

I've always been curious about the performance implications of using `hash` to generate unique keys in certain components. I believe the hashes of data structures are cached in ClojureScript, so the performance could be surprisingly good. Or maybe not?

Anyway, in this case the timestamp + index is more than enough.

Reply:
I had a quick look at it, just out of curiosity, https://github.com/clojure/clojurescript/blob/e7cdc70d0371a26e07e394ea9cd72d5c43e5e363/src/main/cljs/cljs/core.cljs#L1002-L1036 

overall seems like it's a potentially slow method, since it does quite a lot, albeit some paths are probably quick.


This is confirmed by rough performance checks:

```
(time (doseq [i (range 0 1000000)] (hash timestamp)))
"Elapsed time: 958.921829 msecs"
```

```
(time (doseq [i (range 0 1000000)] (str "something" timestamp i)))
"Elapsed time: 609.949371 msecs"
```

Results are fairly consistent (900ms vs 600ms).

But you know what's much faster ;) ?

```
 ^{:key i}
 ```
(no price other than allocation of the map I guess)

As far as I know, that works fine, keys need only to be unique among siblings, if that works, is by far better, (string concatenation is expensive, without it, it shaves off 450ms from the running time of `(str..`

Reply:
@alwx if that works and does not complain, would you mind changing the `key` to just the index please?

Reply:
mmh, me dummy, forget that, of course it doesn't work on updates/changes of the array, apologies

Reply:
don't we have a notification id @ilmotta there? so that could just be `^{:key id}`

Reply:
oh nevermind, `{:key i}` works just as well as `{:key (str "..." timestamp)}`, so we should just use `{:key i}` in this case, as per first comment, sorry for the spam

Reply:
Last comment I promise :)

So, neither approach are correct, I think:

`{:key (str timestamp i)}` is a worse version of `{:key i}`, since timestamp is the same for all the siblings, and should add no benefits. 

A notification can change state, in which case items will be re-rendered, but neither key scheme addresses this, since the key will not change for some elements, but the `items` will be different (i.e different buttons with the same id, in both scheme), the assumption is that `timestamp` does not change with a notification state change.

Simplest way, I would just add a `key` in each `items` (it needs to be unique per combination of state+action)

```
:items     (case membership-status
                   constants/activity-center-membership-status-accepted
                   [{:type    :status
                     :subtype :positive
                     :key :activity-center-membership-accepted
                     :label   (i18n/label :t/accepted)}]

                   constants/activity-center-membership-status-declined
                   [{:type    :status
                     :subtype :negative
                      :key :activity-center-membership-declined
                     :label   (i18n/label :t/declined)}]
 

            ...
```

We can also use accessibility-id in theory, but we'd be conflating two things.

I think that's the safest solution, does it make sense?

Reply:
>I had a quick look at it, just out of curiosity, https://github.com/clojure/clojurescript/blob/e7cdc70d0371a26e07e394ea9cd72d5c43e5e363/src/main/cljs/cljs/core.cljs#L1002-L1036

Indeed, I looked up the implementation and double-checked how Clojure does hashing and equality checks. There's always something new to learn.

>mmh, me dummy, forget that, of course it doesn't work on updates/changes of the array, apologies

But it's surprisingly common to see code using just the index as key. I see some suspicious usages in the mobile codebase.

>Simplest way, I would just add a key in each items (it needs to be unique per combination of state+action)

Excellent suggestion.